Add Solidity test case generator framework

bshastry opened the pull request:
~~Fixes #9676~~

bshastry committed:
Add solidity generator

Co-authored-by: Kamil Śliwak <kamil.sliwak@codepoets.it>

cameel commented on test/tools/ossfuzz/SolCustomMutator.h:
```diff
@@ -0,0 +1,43 @@
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
```
Please add a short docstring that describes the whole file (here and in the other modules added in this PR).

cameel commented on test/tools/ossfuzz/SolCustomMutator.h:
```diff
@@ -0,0 +1,43 @@
+	std::string In{};
+	std::string Out{};
+	size_t MaxMutantSize = 0;
+	unsigned int Seed = 0;
```
Is there anything in the fuzzer that requires these to be named in some specific way? If not then I think these should follow our naming convention for plain structs (i.e. start with a lower case letter).

```suggestion
	uint8_t* data = nullptr;
	size_t size = 0;
	std::string in{};
	std::string out{};
	size_t maxMutantSize = 0;
	unsigned int seed = 0;
```

cameel replied:
There's no point setting defaults on anything except `Out` since the only constructor requires specifying all the values anyway. And if `Out` is the only one, I'd set it on the constructor initializer list along with the other values.

cameel commented on test/tools/ossfuzz/SolCustomMutator.cpp:
```diff
@@ -0,0 +1,87 @@
+	In = string(_data, _data + _size),
+	MaxMutantSize = _maxSize;
+	Seed = _seed;
+}
```
Why not use constructor initializer list?
```suggestion
SolCustomMutator::SolCustomMutator(uint8_t* _data, size_t _size, size_t _maxSize, unsigned int _seed):
    Data(_data),
    Size(_size),
    In(string(_data, _data + _size)),
    MaxMutantSize(_maxSize),
    Seed(_seed)
{}
```

cameel commented on test/tools/ossfuzz/SolCustomMutator.cpp:
```diff
@@ -0,0 +1,87 @@
+		Tokens.fill();
+		SolidityParser Parser(&Tokens);
+		Parser.removeErrorListeners();
+		SolidityMutator Mutator(Seed);
```
These local variable names should also start in lower case.

cameel commented on test/tools/ossfuzz/SolCustomMutator.cpp:
```diff
@@ -0,0 +1,87 @@
+			return LLVMFuzzerMutate(Data, Size, MaxMutantSize);
+		else
+		{
+			size_t mutantSize = Out.size() >= MaxMutantSize ? MaxMutantSize - 1 : Out.size();
```
```suggestion
			size_t mutantSize = min(MaxMutantSize - 1, Out.size());
```

cameel commented on test/tools/ossfuzz/SolCustomMutator.cpp:
```diff
@@ -0,0 +1,87 @@
+	{
+		// Default to libFuzzer's mutator
+		return LLVMFuzzerMutate(Data, Size, MaxMutantSize);
+	}
```
I'd put the handler over the smallest scope possible, not to silence this exception being thrown from other places for unrelated reasons. I.e. only around `antlr4::ANTLRInputStream AStream(In);`.

cameel commented on test/tools/ossfuzz/SolCustomMutator.cpp:
```diff
@@ -0,0 +1,87 @@
+)
+{
+	if (_maxSize <= _size || _size == 0)
+		return LLVMFuzzerMutate(_data, _size, _maxSize);
```
Does this actually happen under normal circumstances? If not then I think an assertion would be more appropriate.

cameel commented on test/tools/ossfuzz/SolCustomMutator.cpp:
```diff
@@ -0,0 +1,87 @@
+
+SolCustomMutator::SolCustomMutator(uint8_t* _data, size_t _size, size_t _maxSize, unsigned int _seed)
+{
+	Data = _data;
```
`mutate()` will do `memcpy()` to address zero if `_data` is `nullptr`. I think you should have an assertion against that here.

cameel replied:
I see you added the assertion but I'd put it at the beginning of the function (or even better - in the constructor, where it's initialized) because now it still runs after stuff like `string(data, data + size)`.

cameel commented on test/tools/ossfuzz/SolCustomMutator.h:
```diff
@@ -0,0 +1,43 @@
+
+	uint8_t* Data = nullptr;
+	size_t Size = 0;
+	std::string In{};
```
`In` is not used at all in the implementation. It should probably be removed.

cameel commented on test/tools/ossfuzz/SolidityMutator.h:
```diff
@@ -0,0 +1,76 @@
+
+namespace solidity::test::fuzzer {
+	using RandomEngine = std::mt19937_64;
+	class SolidityMutator: public SolidityBaseVisitor
```
We don't usually indent `namespace` content. Even the other files in this PR don't.

cameel commented on test/tools/ossfuzz/SolidityMutator.h:
```diff
@@ -0,0 +1,76 @@
+		/// Mutant stream
+		std::ostringstream Out;
+		/// Random number generator
+		RandomEngine Rand;
```
```suggestion
		/// Mutant stream
		std::ostringstream m_out;
		/// Random number generator
		RandomEngine m_rand;
```

This is actually more annoying than I'd expect it to be. Took me too much time to realize that `Rand` is just a member and not some class included from somewhere else :)

cameel commented on test/tools/ossfuzz/SolidityMutator.h:
```diff
@@ -0,0 +1,76 @@
+		/// @returns a pseudo randomly generated string of length @param _length
+		/// characters where each character in the string is in the ASCII character
+		/// range between @param _start and @param _end characters.
+		std::string genRandString(unsigned _length, char _start = '!', char _end = '~');
```
I think that the code reads better with full names. Our coding style also recommends not shortening anything except for extremely common stuff.
```suggestion
		std::string generateRandomString(unsigned _length, char _start = '!', char _end = '~');
```

cameel commented on test/tools/ossfuzz/SolidityMutator.h:
```diff
@@ -0,0 +1,76 @@
+		unsigned randOneToN(unsigned _n = 20)
+		{
+			return Rand() % _n + 1;
+		}
```
I think that the results from `mt19937_64` passed through `%` might not necessarily be uniformly distributed, especially for very large `_n`. I'd use `std::uniform_int_distribution` here just to be safe.

cameel replied:
I'd assert that `_n != 0`.

cameel commented on test/tools/ossfuzz/SolidityMutator.cpp:
```diff
@@ -0,0 +1,125 @@
+	else
+	{
+		// Generate a random pragma that does not contain a semicolon
+		pragma = genRandString(randOneToN(30), '<');
```
It would be a bit more readable without relying on the default. Otherwise it's easy to assume that it will just generate a random number of `<` characters.

Also, if `<` was chosen as the first ASCII character after the semicolon then I think that `';' + 1` would be more self-explanatory.
```suggestion
		pragma = genRandString(randOneToN(30), ';' + 1, '~');
```

cameel commented on test/tools/ossfuzz/SolidityMutator.cpp:
```diff
@@ -0,0 +1,125 @@
+
+antlrcpp::Any SolidityMutator::visitPragmaDirective(SolidityParser::PragmaDirectiveContext*)
+{
+	auto PickLiteral = [](unsigned const _idx) -> string
```
```suggestion
	auto PickPragmaLiteral = [](unsigned const _idx) -> string
```
It really only generates literals for use in the pragma, not just any random literal.

cameel commented on test/tools/ossfuzz/SolidityMutator.cpp:
```diff
@@ -0,0 +1,125 @@
+	{
+		// Generate a random pragma that does not contain a semicolon
+		pragma = genRandString(randOneToN(30), '<');
+		pragma.erase(remove(pragma.begin(), pragma.end(), ';'), pragma.end());
```
This seems superfluous given that the line above it generates random strings not containing semicolons.

cameel commented on test/tools/ossfuzz/SolidityMutator.cpp:
```diff
@@ -0,0 +1,125 @@
+using namespace solidity::test::fuzzer;
+
+string SolidityMutator::genRandString(unsigned int _length, char _start, char _end)
+{
```
It won't work properly unless `_start` <= `_end`, right?
```suggestion
string SolidityMutator::genRandString(unsigned int _length, char _start, char _end)
{
	assert(_start <= _end);
```

I think it's really easy to make that mistake when it comes to symbols and it would be useful to be able to detect such mistakes.

cameel commented on test/tools/ossfuzz/SolidityMutator.cpp:
```diff
@@ -0,0 +1,125 @@
+			}
+		);
+		Out << Whiskers(R"(import {<syms>} from <path>;<nl>)")
+			("syms", boost::algorithm::join(symbols, ", "))
```
```suggestion
		Out << Whiskers(R"(import {<symbols>} from <path>;<nl>)")
			("symbols", boost::algorithm::join(symbols, ", "))
```

cameel commented on test/tools/ossfuzz/SolidityMutator.h:
```diff
@@ -0,0 +1,76 @@
+
+		/// @returns a pseudo randomly chosen unsigned number between one
+		/// and @param _n
+		unsigned randOneToN(unsigned _n = 20)
```
I don't think this default it's useful. It just makes it harder to spot the mistake if you forget to specify the number you want.

cameel commented on test/tools/ossfuzz/SolidityMutator.h:
```diff
@@ -0,0 +1,76 @@
+
+#pragma once
+
+#include <test/tools/ossfuzz/SolidityBaseVisitor.h>
```
Is this file autogenerated? I can't find it in the repo but tests pass on this PR so it must be available somehow...

bshastry replied:
Yes. These lines auto generate the antlr visitors the mutator inherits: https://github.com/ethereum/solidity/blob/9ced804236153e5202a6b77b5946cf4840fe526a/scripts/ci/build_ossfuzz.sh#L35-L37

The said lines are invoked by the ossfuzz build CI.

cameel commented on test/tools/ossfuzz/SolidityMutator.cpp:
```diff
@@ -0,0 +1,125 @@
+		pragma.erase(remove(pragma.begin(), pragma.end(), ';'), pragma.end());
+	}
+
+	Out << Whiskers(R"(pragma <string>;<nl>)")
+		("string", pragma)
+		("nl", "\n")
+		.render();
+	return antlrcpp::Any();
```
Does it actually work? I see that the mutated code is returned via `Out` but it's a member variable and not something inherited.

Is this class just used as a template parameter by something in libfuzzer? That would explain having to stick to the weird naming convention...

bshastry replied:
Sorry for the naming convention. No, it is not due to something in the libfuzzer. I suppose instead of a stringstream member variable, I could also have the visitors return a string and keep concatenating but I think that is more of a cosmetic change.

cameel replied:
But what is accessing the stringstream? It's not in the base class so it does not know about it. And there's nothing in this PR that gets the data out of it directly.

cameel replied:
Oh, I see it now. The mutator has `toString()` method and you do call it. Not sure how I missed that :)

cameel commented on test/tools/ossfuzz/SolCustomMutator.h:
```diff
@@ -0,0 +1,43 @@
+
+namespace solidity::test::fuzzer
+{
+struct SolCustomMutator
```
I think this class needs a more self-explanatory name. When I started reading this code I was wondering why it just calls the other mutator without doing anything extra. Now I see that one is a mutator and the other is just part of its implementation.

bshastry replied:
Sorry for the confusion. May be `SolMutatorInterface` is more appropriate here?

bshastry replied:
Changed name.

cameel replied:
It's much better because it does the make the distinction clear. It now sounds like an abstract base class serving as an interface though. Maybe `SolidityCustomMutatorWrapper`?

Or you could rename both clases to something like:
- `SolidityCustomMutatorWrapper` -> `SoliditySourceCodeMutator`
- `SolidityMutator` -> `SolidityASTMutator`

cameel commented on test/tools/ossfuzz/SolCustomMutator.cpp:
```diff
@@ -0,0 +1,87 @@
+		Out = Mutator.toString();
+
+		// If mutant is empty, default to libFuzzer's mutator
+		if (Out.empty())
+			return LLVMFuzzerMutate(Data, Size, MaxMutantSize);
```
Can this actually happen? Maybe an assertion would be more appropriate?

cameel commented on test/tools/ossfuzz/SolCustomMutator.cpp:
```diff
@@ -0,0 +1,87 @@
+			return mutantSize;
+		}
+	}
+	// Range error is thrown by antlr4 runtime's ANTLRInputStream
+	// See https://github.com/antlr/antlr4/issues/2036
+	catch (range_error const&)
+	{
+		// Default to libFuzzer's mutator
+		return LLVMFuzzerMutate(Data, Size, MaxMutantSize);
+	}
```
Is this the right way to handle this? If I understand correctly this means that if the input has invalid UTF-8 characters, the fuzzer will still run and but will just not use the custom mutator. Wouldn't it be better to report an error instead? Or just skip tests with known bad input at a higher level?

bshastry replied:
Good point. Actually the reason I ran into this exception was because antlr assumes an utf-8 encoding and converts it to utf-16 for some reason and the converter throws this exception. Do you have any suggestions to detect invalid utf-8 characters in an opaque memory stream i.e., uint8_t* region?

cameel replied:
I don't think there's anything in STL. Boost regex library has [unicode_iterators](https://www.boost.org/doc/libs/latest/libs/regex/doc/html/boost_regex/ref/internals/uni_iter.html) and maybe it could be used for that?

cameel reviewed (commented):
I have never touched the fuzzer so far so I have a few questions mixed in with the comments. Sorry if some of them are about obvious things :)

The rest are mostly issues/suggestions or assertions but there are a few more concrete ones. I haven't found anything that would be actually broken assuming the tests in the PR actually run.

bshastry commented:
Thank you for the review @cameel . I will slowly work through your feedback and try to make the code cleaner and make it conform to our coding style.

cameel commented on test/tools/ossfuzz/SolidityMutator.h:
```diff
@@ -0,0 +1,68 @@
+	std::ostringstream m_out;
+	/// Random number generator
+	RandomEngine m_rand;
+	/// Whisker template strings for Solidity syntax
+	const std::string s_pragmaDirective = R"(pragma <directive>;)";
```
It's not static (perhaps it was but you changed it to `const`?) so the `s_` prefix should not be there.

cameel commented on test/tools/ossfuzz/SolidityMutator.h:
```diff
@@ -0,0 +1,68 @@
+
+namespace solidity::test::fuzzer {
+using RandomEngine = std::mt19937_64;
+using Dist = std::uniform_int_distribution<uint32_t>;
```
```suggestion
using Distribution = std::uniform_int_distribution<uint32_t>;
```

cameel commented on test/tools/ossfuzz/SolidityCustomMutatorInterface.h:
```diff
@@ -0,0 +1,54 @@
+
+#pragma once
+
+#include <liblangutil/Exceptions.h>
+
+#include <cstddef>
+#include <cstdint>
+#include <ostream>
+#include <random>
```
I think `<random>` is unused in this module.

leonardoalt replied:
I also think `random` can be removed

leonardoalt replied:
Actually all of them

cameel commented on test/tools/ossfuzz/SolidityMutator.h:
```diff
@@ -0,0 +1,68 @@
+
+#include <random>
+
+namespace solidity::test::fuzzer {
```
```suggestion
namespace solidity::test::fuzzer
{
```

cameel commented on test/tools/ossfuzz/SolidityMutator.h:
```diff
@@ -0,0 +1,68 @@
+{
+public:
+	SolidityMutator(unsigned int _seed): m_rand(_seed) {}
+	std::string toString()
```
```suggestion
	std::string toString() const
```

leonardoalt commented:
@bshastry Just pinging

leonardoalt commented:
@bshastry I noticed you pushed a few hours ago, is this ready for review again?

bshastry commented:
Sorry for the late update on this PR @leonardoalt  and @cameel 

There have been some changes, hopefully for the better. 
  - Removed antlr4 dependency since it didn't really serve original use case (see https://github.com/ethereum/solidity/pull/10044)
  - Removed everything from this PR except a basic skeleton for quicker re-review
  - Generator currently only "generates" `pragma solidity >= 0.0.0;` but of course future reviews will incorporate grammar elements found in the Solidity antlr grammar spec. This will be written in C++ with minimal dependencies.

leonardoalt commented on test/tools/ossfuzz/SolidityCustomMutatorInterface.cpp:
```diff
@@ -0,0 +1,67 @@
+		"Solc custom mutator: Invalid mutant or memory pointer"
+	);
+	size_t mutantSize = min(testCase.size(), maxMutantSize - 1);
+	mempcpy(data, testCase.data(), mutantSize);
```
Who's responsible for the allocation of `data`?

bshastry replied:
> Who's responsible for the allocation of `data`?

libFuzzer (https://github.com/llvm-mirror/compiler-rt/blob/69445f095c22aac2388f939bedebf224a6efcdaf/lib/fuzzer/FuzzerInterface.h#L49-L55)

leonardoalt commented on test/tools/ossfuzz/SolidityGenerator.h:
```diff
@@ -0,0 +1,44 @@
+	SolidityGenerator(unsigned int _seed): m_rand(_seed)
+	{}
+	/// @returns a pseudo randomly generated test program
+	std::string visit();
```
Any particular reason why the function is called `visit`? How about `operator()`?

bshastry replied:
Hmm... Yeah, visit is undescriptive. Perhaps `generateTestCase`?

leonardoalt replied:
Sounds good

leonardoalt commented on test/tools/ossfuzz/SolidityGenerator.h:
```diff
@@ -0,0 +1,42 @@
+	std::string generateTestProgram();
+private:
+	/// Random number generator
+	RandomEngine m_rand;
```
Will `m_rand` ever change? Can it be `const`?

bshastry replied:
`m_rand` is seeded here, so it cant be `const` iiuc

https://github.com/ethereum/solidity/pull/9722/files#diff-1b015d56d122f4307a3485f8b0c74e56df64a2da9d5cc28de179fd102d416d66R34

Essentially, each mutation has to be deterministic. To enable this, `libFuzzer` provides a unique seed for each mutation which is passed to the local `RandomEngine` instance.

leonardoalt replied:
Right, but it can still be `const` and initialized in the constructor @bshastry

bshastry replied:
Hmm... If I do `RandomEngine const& m_rand` and initialize `m_rand(_seed)` the compiler complains that I am trying to bind to an lvalue of type `unsigned`. Something does not seem right

leonardoalt replied:
Not a reference though, just
```suggestion
	RandomEngine const m_rand;
```

axic commented on test/tools/ossfuzz/SolidityGenerator.cpp:
```diff
@@ -0,0 +1,32 @@
+using namespace solidity::util;
+using namespace std;
+
+string SolidityGenerator::generateTestProgram()
```
So this PR is establishing the framework? Given this generated starting contract doesn't seem to be very exciting.

leonardoalt replied:
Yea I think that's the idea

bshastry replied:
Yeah, this PR is setting up the framework. Code for Test case generation will follow in upcoming PRs

axic replied:
Can you add a comment here explaining what this does and what needs to be done? And what is blocking this PR?

bshastry replied:
Done

axic replied:
Shouldn't this have an input, like the seed? Or how will this generate code?

bshastry replied:
The idea is to write a generator based on the antlr4 grammar here. https://github.com/ethereum/solidity/blob/develop/docs/grammar/Solidity.g4

Starting point is a test unit that contains one or more source units. Each source unit contains one or more elements such as function, contract etc.

The input to the generator is a deterministic seed (`uint64_t`) passed here

https://github.com/ethereum/solidity/pull/9722/files#diff-3f64bed29063599c885c0e99d8aa0d98734424f0fac0009517c0b04ba0cb0a8fR38

This seed fully determines the sequence of generation steps undertaken by the generator so that a given seed always produces the same test program.

The test program itself is generated by randomly choosing one of N elements from the grammar and recursive descent.

leonardoalt replied:
The seed is given and used in the constructor

leonardoalt commented:
Might also need a rebase

bshastry commented:
Rebased. Not sure why spellcheck failed since its complaints are unrelated to this PR.

leonardoalt commented:
Yea also not sure, it also fails in another PR

axic commented on test/tools/ossfuzz/CMakeLists.txt:
```diff
@@ -31,6 +33,16 @@ if (OSSFUZZ)
+            SolidityGenerator.cpp
+            SolidityCustomMutatorInterface.cpp
+    )
+    target_link_libraries(solc_opt_mutator_ossfuzz PRIVATE libsolc evmasm)
```
Why do we need to link `evmasm` explicitly and why isn't `libsolc` enough?

bshastry replied:
Hmm... Looks like I copied this line from here: https://github.com/ethereum/solidity/blob/aa6e96ce886f87550769dc5cbdec624a49f1fc5b/test/tools/CMakeLists.txt#L8

Without linking in evmasm, fuzzer build complains about not being able to locate `fuzzer_common.h`. Strange.
